### PR TITLE
Step1

### PR DESCRIPTION
**Pull request by pasaunders**
The server returns the expected data and passes all required tests.

### CODE REVIEW

File: src/server.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            msg_body = total_message[1]
+            try:
+                final_message = response_ok() + "".join(msg_body)
+                print(final_message, type(final_message))
+                conn.sendall(final_message.encode('utf8'))
+            except:
+                conn.sendall(response_error().encode('utf8'))
+        except KeyboardInterrupt:
+            break
+    conn.close()
```
Comment (WillWeatherford) [inline src/server.py]: Similar to your echo server, this line right here is causing your "already in use" errors. You're creating a new `conn` inside your while loop everytime a new request comes in, but the call to close is on the outside of the loop.

File: src/server.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    server.listen(1)
+
+    while True:
+        conn, addr = server.accept()
+        try:
+            total_message = ""
+            buffer_length = 1024
+            while total_message.count("\r\n\r\n") != 2:
+                part = conn.recv(buffer_length)
+                total_message += part.decode('utf8')
```
Comment (WillWeatherford) [inline src/server.py:25]: Similar to with your echo server, you're going to have big problems handling unicode chars decoding them one chunk at a time. Also, multiple concatenations is a computationally inefficient way to build a string from chunks. A better approach is to append chunks to a list, then use the str.join method on the list.

File: src/server.py (hunk 1 above)
Comment (rveeblefetzer) [reply src/server.py]: thanks, Will. Addressing this (and the other comments) now

File: src/client.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    stream_info = [i for i in infos if i[1] == socket.SOCK_STREAM][0]
+    client = socket.socket(*stream_info[:3])
+    client.connect(stream_info[-1])
+    client.sendall(send_msg.encode('utf8'))
+    reply = ""
+    buffer_length = 1024
+    reply_complete = False
+    while not reply_complete:
+        part = client.recv(buffer_length)
+        reply += part.decode('utf8')
```
Comment (nhuntwalker) [inline src/client.py:24]: This breaks in Python 2

File: src/server.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    #         except:
+    #             conn.sendall(response_error().encode('utf8'))
+    #             total_message += part
+    #         conn.sendall(total_message)
+    #     except KeyboardInterrupt:
+    #         break
+    #     conn.close()
+    # import pdb; pdb.set_trace()
+    # server.shutdown()
+    # server.close()
```
Comment (nhuntwalker) [inline src/server.py:62]: Corpse code

File: src/server.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    msg_head = total_message[0]
+    print('Message head: ', msg_head)
+    request_bits = msg_head.split()
+    print(request_bits[0])
+    print('GET')
+    print(request_bits[0] == 'GET')
+    print(request_bits[1] == '/index.html')
+    print(request_bits[2] == 'HTTP/1.1')
+    print(request_bits[4] == 'www.example.com')
+    print('request bits: ', request_bits)
```
Comment (nhuntwalker) [inline src/server.py:77]: What's with all these print statements?

File: src/server.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    try:
+        if msg_head == b'GET /index.html HTTP/1.1\r\nHost: www.example.com':
+            return [response_ok(), request_bits[4]]
+        elif request_bits[0] != 'GET':
+            raise ValueError
+        elif request_bits[1] != '/index.html':
+            print('path is wrong:', request_bits[1])
+            raise ValueError
```
Comment (nhuntwalker) [inline src/server.py:85]: You don't need a separate print statement. Include the message as a part of the exception being raised

File: src/server.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        elif request_bits[2] != 'HTTP/1.1':
+            print('HTTP version is wrong: ', request_bits[2])
+            raise ValueError
```
Comment (nhuntwalker) [inline src/server.py:88]: Same as above.

File: src/server.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        elif request_bits[4] != 'www.example.com':
+            print('URI is wrong: ', request_bits[4])
+            raise ValueError
```
Comment (nhuntwalker) [inline src/server.py:91]: Same as above.

File: src/step1.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+"""A simple HTTP server."""
```
Comment (nhuntwalker) [inline src/step1.py:1]: What is this document for?

**Pull request merged**